src/librustc_resolve/diagnostics.rs

```diff
@@ -444,6 +444,52 @@ impl SomeTrait for Foo { // ok!
 ```
 "##,
 
+E0406: r##"
+A function is referring to an associated type which hasn't been declared in
+the trait. Example of erroneous code:
+
+```compile_fail
+trait Foo {
+    type Bar;
+
+    // error: function signature contains a reference to `Self::Baz`, but
+    //        `Baz` hasn't been declared as an associated type of the trait
+    fn return_bool(&self, &Self::Bar, &Self::Baz) -> bool;
+}
+```
+You tried to use an associated type which hasn't been declared in the
+trait body. All associated types must be declared with the
+type keyword, e.g. `type a;` before being used.
```

Comment:
I don't like very much "e.g.". But since I don't have something in mind to replace, let's keep it (except if you find something better).